system_metrics_collector/src/topic_statistics_collector/received_message_age.hpp

```diff
@@ -45,21 +52,28 @@ class ReceivedMessageAgeCollector : public TopicStatisticsCollector<T>
 
   /**
   * Handle a new incoming message. Calculate message age if a valid Header is present.
+  *
+  * @param received_message, the message to calculate age of.
   */
   void OnMessageReceived(const T & received_message) override
   {
-    const rclcpp::Time timestamp_from_header = message_filters::message_traits::TimeStamp<T>::value(
+    const auto timestamp_from_header = message_filters::message_traits::TimeStamp<T>::value(
       received_message);
 
     if (timestamp_from_header.nanoseconds()) {
       const auto now = GetCurrentTime();
 
-      const std::chrono::nanoseconds age_nanos{
-        now.nanoseconds() - timestamp_from_header.nanoseconds()};
-      const std::chrono::milliseconds age_millis{
-        std::chrono::duration_cast<std::chrono::milliseconds>(age_nanos)};
-
-      system_metrics_collector::Collector::AcceptData(static_cast<double>(age_millis.count()));
+      if (timestamp_from_header.get_clock_type() == now.get_clock_type()) {
+        const std::chrono::nanoseconds age_nanos{
+          now.nanoseconds() - timestamp_from_header.nanoseconds()};
+        const auto age_millis = std::chrono::duration_cast<std::chrono::milliseconds>(age_nanos);
+
+        system_metrics_collector::Collector::AcceptData(static_cast<double>(age_millis.count()));
+      } else {
+        RCUTILS_LOG_WARN_NAMED(
+          "topic_statistics_collector",
+          "Message header and current clock have different time sources, no metric reported");
```

Comment:
```suggestion
          "Message header and current clock have different time sources, cannot measure message age");
```

I'd also note the clock types used for the message and this class.